### PR TITLE
Move training op helpers to framework and fix linking issues

### PR DESCRIPTION
**Pull request by samikama**
Hi, this PR encompasses the PR #37873 and fixes the linking issues. @alextp @Squadrick @spidyDev

### CODE REVIEW

**Comment by samikama** [comment]
@vcarpani @Squadrick Since I keep your commits from earlier PR to credit your work, could you please approve the CLA?

**Comment by vcarpani** [comment]
@googlebot I consent.

**Comment by spidyDev** [comment]
@samikama  Thanks for making this change. The PR works fine with custom ops  in TF addons.

**Review by alextp — DISMISSED** [review]

**Comment by rthadur** [comment]
@samikama can you please sign CLA ?

**Comment by samikama** [comment]
@googlebot I consent

**Comment by samikama** [comment]
@Squadrick do you consent to include your commits in PR or should I open a new one without your changes?

**Comment by samikama** [comment]
@googlebot I consent.

**Comment by Squadrick** [comment]
@googlebot I consent.

**Review by alextp — DISMISSED** [review]

**Review by alextp — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Comment by samikama** [comment]
@rthadur, it looks like copybara failed but can't see the reason. Is there anything on my side to help?

**Review by rthadur — DISMISSED** [review]

**Review by alextp — DISMISSED** [review]

**Comment by samikama** [comment]
@rthadur and @alextp There seems to be an issue with copybara but I can't see the error. Is there anything I can do to help?

**Comment by mihaimaruseac** [comment]
Copybara fails because we no longer have `contrib`. Can we remove the `contrib` files from the PR?

**Comment by samikama** [comment]
@mihaimaruseac Thanks for catching that. I thought these were removed in earlier PRs that this is based on. I removed contrib directory now.

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by samikama** [comment]
@mihaimaruseac I believe current failures are not related with this PR, since there wasn't any change that would have caused them wrt previous commit. Is that correct?

**Comment by mihaimaruseac** [comment]
The MacOS failures are not related. They also won't block merging this. Currently this is waiting import approval, but it shouldn't take longer.

File: tensorflow/core/kernels/stateful_random_ops.cc [hunk 1]
```diff
@@ -16,9 +16,9 @@ limitations under the License.
 #define EIGEN_USE_THREADS
 
 #include "tensorflow/core/framework/tensor_util.h"
+#include "tensorflow/core/framework/training_op_helpers.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/stateful_random_ops.cc:19]: `tensorflow/core/kernels:stateful_random_ops does not depend on a module exporting 'third_party/tensorflow/core/framework/training_op_helpers.h'`

File: tensorflow/core/kernels/resource_variable_ops.cc [hunk 2]
```diff
@@ -62,15 +62,15 @@ limitations under the License.
 #include "tensorflow/core/framework/op_kernel.h"
 #include "tensorflow/core/framework/register_types.h"
 #include "tensorflow/core/framework/resource_mgr.h"
+#include "tensorflow/core/framework/resource_var.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/resource_variable_ops.cc:65]: `tensorflow/core/kernels:resource_variable_ops does not depend on a module exporting 'third_party/tensorflow/core/framework/resource_var.h'`

File: tensorflow/core/kernels/resource_variable_ops.cc [hunk 3]
```diff
@@ -62,15 +62,15 @@ limitations under the License.
...
 #include "tensorflow/core/framework/tensor_shape.h"
 #include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/framework/training_op_helpers.h"
 #include "tensorflow/core/framework/variant_op_registry.h"
 #include "tensorflow/core/kernels/dense_update_functor.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/resource_variable_ops.cc:70]: `tensorflow/core/kernels:resource_variable_ops does not depend on a module exporting 'third_party/tensorflow/core/kernels/dense_update_functor.h'`

File: tensorflow/core/kernels/count_up_to_op.cc [hunk 4]
```diff
@@ -14,7 +14,8 @@ limitations under the License.
 ==============================================================================*/
 
 #include "tensorflow/core/framework/op_kernel.h"
-#include "tensorflow/core/kernels/variable_ops.h"
+#include "tensorflow/core/framework/resource_mgr.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/count_up_to_op.cc:17]: `/tensorflow/core/kernels/variable_ops.cc:19:10: error: use of private header from outside its module: '/tensorflow/core/framework/op_kernel.h'; 
         ^/tensorflow/core/kernels/variable_ops.cc:20:10: error: use of private header from outside its module: '/tensorflow/core/framework/register_types.h';

File: tensorflow/core/kernels/scatter_nd_op.cc [hunk 5]
```diff
@@ -25,13 +25,14 @@ limitations under the License.
 #include "tensorflow/core/framework/bounds_check.h"
 #include "tensorflow/core/framework/op_kernel.h"
 #include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/resource_mgr.h"
+#include "tensorflow/core/framework/resource_var.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/scatter_nd_op.cc:29]: `/tensorflow/core/kernels/scatter_nd_op.cc:29:10: error: module //
/tensorflow/core/kernels:scatter_nd_op does not depend on a module exporting '/tensorflow/core/framework/resource_var.h'

/tensorflow/core/kernels/scatter_nd_op.cc:33:10: error: module //
/tensorflow/core/kernels:scatter_nd_op does not depend on a module exporting '/tensorflow/core/kernels/dense_update_functor.h'`

File: tensorflow/core/kernels/strided_slice_op.cc [hunk 6]
```diff
@@ -28,14 +28,13 @@ limitations under the License.
 #include "tensorflow/core/framework/bounds_check.h"
 #include "tensorflow/core/framework/op_kernel.h"
 #include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/resource_var.h"
 #include "tensorflow/core/framework/tensor.h"
-#include "tensorflow/core/kernels/dense_update_functor.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/strided_slice_op.cc:32]: `tensorflow/core/kernels/strided_slice_op.cc:29:10: error: module //
/tensorflow/core/kernels:strided_slice_op does not depend on a module exporting '/tensorflow/core/framework/resource_var.h`

File: tensorflow/core/kernels/training_ops.cc [hunk 7]
```diff
@@ -24,8 +24,8 @@ limitations under the License.
 #include "tensorflow/core/framework/bounds_check.h"
 #include "tensorflow/core/framework/op_kernel.h"
 #include "tensorflow/core/framework/register_types.h"
-#include "tensorflow/core/kernels/training_op_helpers.h"
```
Comment (rthadur) [inline tensorflow/core/kernels/training_ops.cc:27]: `tensorflow/core/framework/training_op_helpers.cc:18:10: error: module //
/tensorflow/core:training_op_helpers does not depend on a module exporting '
/tensorflow/core/util/ptr_util.h`

**Review by rthadur — CHANGES_REQUESTED** [review]
@samikama can you please check below errors ?

**Comment by samikama** [comment]
@rthadur I just did a clean build and can't reproduce the issue. I am using bazel 2.0.0? Is this observed in internal build? If not can you tell me how can I reproduce it?

**Comment by samikama** [comment]
@rthadur any help on this? If I can't see the errors I can't know what is wrong or verify if they are fixed.

**Comment by mihaimaruseac** [comment]
These seem to be caused by `BUILD` files not mirroring the include hierarchy. I'll try to import and fix manually later today/this week

**Comment by mihaimaruseac** [comment]
Unfortunately I have to give up trying to manually import. Errors are either of this form:

```
...$ bazel build //tensorflow/core/kernels:variable_ops
...
ERROR: .../tensorflow/core/kernels/BUILD:5716:18: C++ compilation of rule '//tensorflow/core/kernels:variable_ops' failed (Exit 1)[..] 
tensorflow/core/kernels/variable_ops.cc:19:10: error: use of private header from outside its module: 'tensorflow/core/framework/op_kernel.h'. [-Wprivate-header]
#include "tensorflow/core/framework/op_kernel.h"
         ^
...
FAILED: Build did NOT complete successfully
```

or errors of the form

```
...$ bazel build //tensorflow/core/kernels:count_up_to_op  
...                                                                                                                                                                                                       
ERROR: .../tensorflow/core/kernels/BUILD:5640:18: C++ compilation of rule '//tensorflow/core/kernels:count_up_to_op' failed (Exit 1)[..]                                                                              
tensorflow/core/kernels/count_up_to_op.cc:18:10: error: module //tensorflow/core/kernels:count_up_to_op does not depend on a module exporting 'tensorflow/core/framework/resource_var.h'                                                              
#include "third_party/tensorflow/core/framework/resource_var.h"                                                                    
         ^         
...                           
FAILED: Build did NOT complete successfully  
```

This results in over 16000 tests failing internally.

Can we try splitting this PR into smaller components and merging them one by one. Can assign me as reviewer for each component.

**Comment by samikama** [comment]
Thanks for looking into that @mihaimaruseac. I can try to give another go if I can have a way to reproduce the problem. I believe these issues are originating from an incomplete refactoring in the past. If is it possible to have a small reproducer, it might be easier to solve these issues. Could you provide a simple example manifesting the errors?

**Comment by mihaimaruseac** [comment]
Sure. The errors I got were reproduceable when building the equivalent of `bazel build //tensorflow/core/kernels:variable_ops` and `bazel build //tensorflow/core/kernels:count_up_to_op`,

If you cannot reproduce, can we try a different approach where we duplicate the code first (one PR), slowly migrate targets to use the new code (multiple PRs), then we do one more PR to sync the duplicates and remove the old copy? I could attempt to do that internally but not in the next 1-2 weeks :(

**Comment by samikama** [comment]
@mihaimaruseac, thanks. I just did another clean build and the targets you mention. They all went through without any problems. I used bazel 2.0.0 as well as 3.2.0. I wonder what is so different between imported code and OSS. Could it be an import issue?

**Comment by samikama** [comment]
Also I extended the visibility of the framework header exports to core/kernels. Perhaps this helps

**Comment by mihaimaruseac** [comment]
Let's try a new pull with the visibility extended.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by rthadur** [comment]
@mihaimaruseac there are still internal errors , can you please check.

**Comment by gbaned** [comment]
@samikama Can you please check @rthadur's comments and keep us posted. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by samikama** [comment]
@gbaned This is a bit hard to resolve since we can not reproduce the issue with public TF release and bazel.

**Comment by rthadur** [comment]
@mihaimaruseac can you please with internal errors ?

**Comment by mihaimaruseac** [comment]
I will need to find some uninterrupted time without any urgent priority. Unfortunately this has not been the case yet, sorry about this.

I have the PR on my list and will surely get to fixing it.

There is a merge conflict, can we fix it?

**Comment by gbaned** [comment]
@samikama Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@samikama, Any update on this PR? Please. Thanks!

**Comment by samikama** [comment]
I will rework this PR when I have time.

**Comment by gbaned** [comment]
@samikama  Any update on this PR? and please resolve conflicts. Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**